### PR TITLE
feat : 문제 세트 구현

### PR DESCRIPTION
**Pull request by kshired**
### Issue Number

close: N/A

### 작업 내역

문제 세트 구현

- [x] 문제 세트 엔티티 구현
- [x] 문제 세트 조회, 생성, 업데이트 구현 

### 변경사항

- 의존성 목록 : N/A

### 작업 유형

- [x] 신규 기능 추가
- [ ] 버그 수정
- [ ] 리펙토링
- [ ] 문서 업데이트

### PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점

- 30분만에 뚝딱 만든거라 빠진게 있을 수도 있으니 깊게 봐주세용

### CODE REVIEW

File: src/main/kotlin/io/csbroker/apiserver/controller/v1/AdminController.kt [hunk 1]
```diff
@@ -46,7 +48,8 @@ class AdminController(
     private val problemService: ProblemService,
     private val userAnswerService: UserAnswerService,
     private val userService: UserService,
-    private val notificationService: NotificationService
+    private val notificationService: NotificationService,
+    private val problemSetService: ProblemSetService
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/controller/v1/AdminController.kt:52]: 🚫 **[ktlint]** <trailing-comma-on-declaration-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/controller/v1/AdminController.kt [hunk 2]
```diff
@@ -325,4 +328,19 @@ class AdminController(
             )
         )
     }
+
+    @PostMapping("/problem-sets")
+    fun createProblemSet(
+        @RequestBody problemSetUpsertRequestDto: ProblemSetUpsertRequestDto
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/controller/v1/AdminController.kt:334]: 🚫 **[ktlint]** <trailing-comma-on-declaration-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/controller/v1/AdminController.kt [hunk 3]
```diff
@@ -325,4 +328,19 @@ class AdminController(
...
+    ): ApiResponse<Long> {
+        return ApiResponse.success(problemSetService.createProblemSet(problemSetUpsertRequestDto))
+    }
+
+    @PostMapping("/problem-sets/{id}")
+    fun updateProblemSet(
+        @PathVariable("id") id: Long,
+        @RequestBody problemSetUpsertRequestDto: ProblemSetUpsertRequestDto
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/controller/v1/AdminController.kt:342]: 🚫 **[ktlint]** <trailing-comma-on-declaration-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/controller/v1/ProblemSetController.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import io.csbroker.apiserver.service.ProblemSetService
+import org.springframework.web.bind.annotation.GetMapping
+import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.RestController
+
+@RestController
+@RequestMapping("/api/v1/problem-sets")
+class ProblemSetController(
+    private val problemSetService: ProblemSetService
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/controller/v1/ProblemSetController.kt:15]: 🚫 **[ktlint]** <trailing-comma-on-declaration-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/controller/v1/ProblemSetController.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+) {
+    @GetMapping("/{id}")
+    fun findById(
+        @PathVariable("id") id: Long
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/controller/v1/ProblemSetController.kt:19]: 🚫 **[ktlint]** <trailing-comma-on-declaration-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/dto/problem/problemset/ProblemSetDetailResponseDto.kt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package io.csbroker.apiserver.dto.problem.problemset
+
+import io.csbroker.apiserver.dto.problem.ProblemResponseDto
+
+data class ProblemSetDetailResponseDto(
+    val id: Long,
+    val problems: List<ProblemResponseDto>,
+    val name: String,
+    val description: String
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/dto/problem/problemset/ProblemSetDetailResponseDto.kt:9]: 🚫 **[ktlint]** <trailing-comma-on-declaration-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/dto/problem/problemset/ProblemSetResponseDto.kt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package io.csbroker.apiserver.dto.problem.problemset
+
+data class ProblemSetResponseDto(
+    val id: Long,
+    val problemCnt: Int,
+    val name: String,
+    val description: String
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/dto/problem/problemset/ProblemSetResponseDto.kt:7]: 🚫 **[ktlint]** <trailing-comma-on-declaration-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/dto/problem/problemset/ProblemSetUpsertRequestDto.kt [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package io.csbroker.apiserver.dto.problem.problemset
+
+data class ProblemSetUpsertRequestDto(
+    val problemIds: List<Long>,
+    val name: String,
+    val description: String
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/dto/problem/problemset/ProblemSetUpsertRequestDto.kt:6]: 🚫 **[ktlint]** <trailing-comma-on-declaration-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/model/ProblemSet.kt [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+    @OneToMany
+    val problems: MutableList<Problem> = mutableListOf(),
+) {
+    fun toProblemSetResponseDto(): ProblemSetResponseDto {
+        return ProblemSetResponseDto(
+            id!!,
+            problems.size,
+            name,
+            description
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/model/ProblemSet.kt:36]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/model/ProblemSet.kt [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        )
+    }
+
+    fun toProblemSetDetailResponseDto(): ProblemSetDetailResponseDto {
+        return ProblemSetDetailResponseDto(
+            id!!,
+            problems.map { it.toProblemResponseDto(GradingHistoryStats.toGradingHistoryStats(it.gradingHistory)) },
+            name,
+            description
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/model/ProblemSet.kt:45]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+import io.csbroker.apiserver.repository.ProblemSetRepository
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+
+@Service
+@Transactional(readOnly = true)
+class ProblemSetServiceImpl(
+    private val problemSetRepository: ProblemSetRepository,
+    private val problemRepository: ProblemRepository
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt:16]: 🚫 **[ktlint]** <trailing-comma-on-declaration-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @Transactional
+    override fun createProblemSet(problemSetUpsertRequestDto: ProblemSetUpsertRequestDto): Long {
+        val (problemIds, name, description) = problemSetUpsertRequestDto
+        val problems = problemRepository.findAllById(problemIds)
+
+        if (problems.size != problemIds.size) {
+            throw EntityNotFoundException(
+                "[ids = ${problemIds.joinToString(",")}] 에 해당하는 문제 세트를 찾을 수 없습니다."
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt:34]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+            )
+        }
+
+        val problemSet = ProblemSet(
+            name = name,
+            description = description,
+            problems = problems
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt:41]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+    @Transactional
+    override fun updateProblemSet(id: Long, problemSetUpsertRequestDto: ProblemSetUpsertRequestDto): Long {
+        val (problemIds, name, description) = problemSetUpsertRequestDto
+        val problemSet = findById(id)
+        val problems = problemRepository.findAllById(problemIds)
+
+        if (problems.size != problemIds.size) {
+            throw EntityNotFoundException(
+                "[ids = ${problemIds.joinToString(",")}] 에 해당하는 문제 세트를 찾을 수 없습니다."
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt:55]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/model/Problem.kt [hunk 15]
```diff
@@ -53,7 +53,11 @@ abstract class Problem(
     val problemTags: MutableSet<ProblemTag> = mutableSetOf(),
 
     @OneToMany(mappedBy = "problem", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
-    val gradingHistory: MutableList<GradingHistory> = mutableListOf()
+    val gradingHistory: MutableList<GradingHistory> = mutableListOf(),
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "problem_set_id", nullable = true)
+    var problemSet: ProblemSet? = null,
```
Comment (ekzm8523) [inline src/main/kotlin/io/csbroker/apiserver/model/Problem.kt:60]: 분명 고민을 해보셨을 것 같아요. many to many가 아닌 one to many로 가신 이유가 있을까요?
개인적으로는 하나의 문제는 무조건 하나의 문제셋에만 종속되면 유연함이 떨어지지 않을까 라는 생각이 들기는 합니다!

File: src/main/kotlin/io/csbroker/apiserver/model/ProblemSet.kt [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+            name,
+            description,
+        )
+    }
+
+    fun update(name: String, description: String, problems: List<Problem>) {
+        this.name = name
+        this.description = description
+        this.problems.clear()
+        problems.forEach { addProblem(it) }
```
Comment (ekzm8523) [inline src/main/kotlin/io/csbroker/apiserver/model/ProblemSet.kt:53]: 질문 1) this.problems = problems 와 뭐가 다른건가요?

질문 2) addProblem을 보니 problem.problemSet = this로 명시해주던데 여기서는 problem들의 problemSet을 null로 안바꿔주어도 되나요?

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+            )
+        }
+
+        val problemSet = ProblemSet(
+            name = name,
+            description = description,
+            problems = problems,
+        )
+
+        return problemSetRepository.save(problemSet).id!!
```
Comment (ekzm8523) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt:44]: 만약 DB 내부적인 규약에 걸려서 저장이 실패한 경우 exception 처리는 어떻게 되나요?

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+    @Transactional
+    override fun createProblemSet(problemSetUpsertRequestDto: ProblemSetUpsertRequestDto): Long {
+        val (problemIds, name, description) = problemSetUpsertRequestDto
+        val problems = problemRepository.findAllById(problemIds)
+
+        if (problems.size != problemIds.size) {
+            throw EntityNotFoundException(
+                "[ids = ${problemIds.joinToString(",")}] 에 해당하는 문제 목록을 찾을 수 없습니다.",
+            )
+        }
```
Comment (ekzm8523) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt:36]: 이 부분 코드가 중복되네요!! 메서드 추출을 하는게 어떨까요?

**Review by ekzm8523 — COMMENTED** [review]
코멘트 확인해주세요 🙃

File: src/main/kotlin/io/csbroker/apiserver/model/Problem.kt (hunk 15 above)
Comment (kshired) [reply src/main/kotlin/io/csbroker/apiserver/model/Problem.kt:60]: @ekzm8523 어.. 아뇨 지금 깨달았네요 반영해볼게요

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt (hunk 17 above)
Comment (kshired) [reply src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt:44]: save는 보통 실패하는 것을 가정하지 않아요.  (어차피 DB 장애가 아닌이상 안나올 듯.)

size 조정이 필요하면, requestbody 쪽에 제약을 걸어볼게요~

File: src/main/kotlin/io/csbroker/apiserver/model/ProblemSet.kt (hunk 16 above)
Comment (kshired) [reply src/main/kotlin/io/csbroker/apiserver/model/ProblemSet.kt:53]: 1. val ( const ) 라서 바로 못바꿔서, clear 하고 추가하는 로직입니다!
2. 빼먹은거 같아요~ 반영할게요!

File: src/main/kotlin/io/csbroker/apiserver/model/ProblemSet.kt [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    val id: Long? = null,
+
+    @Column(name = "problem_set_name")
+    var name: String,
+
+    @Column(name = "problem_set_description")
+    var description: String,
+
+    @OneToMany
+    val problemSetMapping: List<ProblemSetMapping> = mutableListOf()
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/model/ProblemSet.kt:29]: 🚫 **[ktlint]** <trailing-comma-on-declaration-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+
+@Service
+@Transactional(readOnly = true)
+class ProblemSetServiceImpl(
+    private val problemSetRepository: ProblemSetRepository,
+    private val problemRepository: ProblemRepository,
+    private val problemSetMappingRepository: ProblemSetMappingRepository
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt:19]: 🚫 **[ktlint]** <trailing-comma-on-declaration-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+        val problemSet = findById(id)
+        problemSet.updateContents(problemSetUpsertRequestDto.name, problemSetUpsertRequestDto.description)
+        problemSetMappingRepository.deleteAllByProblemSetId(problemSet.id!!)
+        createProblemSetMapping(problemSetUpsertRequestDto.problemIds, problemSet)
+        return problemSet.id!!
+    }
+
+    private fun createProblemSetMapping(
+        problemIds: List<Long>,
+        problemSet: ProblemSet
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt:48]: 🚫 **[ktlint]** <trailing-comma-on-declaration-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+        if (problems.size != problemIds.size) {
+            throw EntityNotFoundException(
+                "[ids = ${problemIds.joinToString(",")}] 에 해당하는 문제 목록을 찾을 수 없습니다.",
+            )
+        }
+
+        val problemSetMappings = problems.map {
+            ProblemSetMapping(
+                problem = it,
+                problemSet = problemSet
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemSetServiceImpl.kt:60]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

**Review by ekzm8523 — APPROVED** [review]
고생하셨습니다~

**Pull request merged**